### ISSUE TITLE
逻辑回归模型的评估指标——程序清单中TP、FP、FN、TN的计算的疑惑

### ISSUE THREAD

**Issue by Phil-Kuo** [GenTang/intro_ds#3]
唐老师，
在阅读到程序清单5-7时，我有个疑惑：
```
def evaluation(re):
    """
    计算预测结果的查准查全率以及f1
    参数
    ----
    re ：DataFrame，预测结果，里面包含两列：真实值‘lable_code’、预测值‘pred’
    """
    bins = np.array([0, 0.5, 1])
    label = re["label_code"]
    pred = re["pred"]
    tp, fp, fn, tn = np.histogram2d(label, pred, bins=bins)[0].flatten()
    precision = tp / (tp + fp)  # 0.951
    recall = tp / (tp + fn)  # 0.826
    f1 = 2 * precision * recall / (precision + recall)  # 0.884
    print("查准率: %.3f, 查全率: %.3f, f1: %.3f" % (precision, recall, f1))
```
` tp, fp, fn, tn = np.histogram2d(label, pred, bins=bins)[0].flatten()`

我查阅了[np.histogram2d](https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram2d.html?highlight=histogram2d#numpy.histogram2d)的帮助文档，我觉得返回的顺序应当是tn,fp,fn,tp呀？

![image](https://user-images.githubusercontent.com/4710305/47733770-ee43c980-dca3-11e8-82ef-d0aeec618bef.png)

这是我认为返回的H：
![image](https://user-images.githubusercontent.com/4710305/47734666-b63d8600-dca5-11e8-9fc5-a604c05c967d.png)
H[0][0]对应TN，H[0][1]对应FP，H[1][0]对应FN，H[1][1]对应TP。

我也用一个小例子试了一下（这个例子里的tp=3, fp=2, fn=1, tn=1）
```
bins=np.array([0, 0.5, 1])
x=pd.Series([0, 1, 0, 1, 1, 1, 0])
y=pd.Series([1, 1, 0, 0, 1, 1, 1])
a, b, c, d = np.histogram2d(x,y,bins=bins)[0].flatten()
```
运行结果如下：
```
In [119]: a,b,c,d
Out[119]: (1.0, 2.0, 1.0, 3.0)
```
我不知道是否np.histogram2d的帮助文档没有理解到位？还请唐老师指点一二。

**Comment by GenTang** [comment]
我验算了一下，是的。我错了。你的理解是对的。
对于这个计算，其实更好的方法是使用sklearn的，confusion_matrix。具体的代码如下

```Python
from sklearn.metrics import confusion_matrix

label = pd.Series([0, 1, 0, 1, 1, 1, 0])
preds = pd.Series([1, 1, 0, 0, 1, 1, 1])
tn, fp, fn, tp = confusion_matrix(label, preds).ravel()
```

**Comment by GenTang** [comment]
我稍后会将代码更新。谢谢你的提交
对错误感到十分抱歉。

**Comment by Phil-Kuo** [comment]
原来如此，这种方法确实更直观！受教了~

**Comment by GenTang** [comment]
bug已修复，再次感谢提交。
